src/index/DeltaTriples.cpp

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 - 2024, University of Freiburg
+//  Chair of Algorithms and Data Structures.
+//  Authors:
+//    2023 Hannah Bast <bast@cs.uni-freiburg.de>
+//    2024 Julian Mundhahs <mundhahj@tf.uni-freiburg.de>
+
+#include "index/DeltaTriples.h"
+
+#include "absl/strings/str_cat.h"
+#include "index/Index.h"
+#include "index/IndexImpl.h"
+#include "index/LocatedTriples.h"
+#include "parser/TurtleParser.h"
+
+// ____________________________________________________________________________
+LocatedTriples::iterator& DeltaTriples::LocatedTripleHandles::forPermutation(
+    Permutation::Enum permutation) {
+  switch (permutation) {
+    case Permutation::PSO:
+      return forPSO_;
+    case Permutation::POS:
+      return forPOS_;
+    case Permutation::SPO:
+      return forSPO_;
+    case Permutation::SOP:
+      return forSOP_;
+    case Permutation::OSP:
+      return forOSP_;
+    case Permutation::OPS:
+      return forOPS_;
+    default:
+      AD_FAIL();
+  }
+}
+
+// ____________________________________________________________________________
+void DeltaTriples::clear() {
+  triplesInserted_.clear();
+  triplesDeleted_.clear();
+  locatedTriplesPerBlockInPSO_.clear();
+  locatedTriplesPerBlockInPOS_.clear();
+  locatedTriplesPerBlockInSPO_.clear();
+  locatedTriplesPerBlockInSOP_.clear();
+  locatedTriplesPerBlockInOSP_.clear();
+  locatedTriplesPerBlockInOPS_.clear();
+}
+
+// ____________________________________________________________________________
+std::vector<DeltaTriples::LocatedTripleHandles>
+DeltaTriples::locateAndAddTriples(
+    ad_utility::SharedCancellationHandle cancellationHandle,
+    std::span<const IdTriple<0>> idTriples, bool shouldExist) {
+  ad_utility::HashMap<Permutation::Enum, std::vector<LocatedTriples::iterator>>
+      intermediateHandles;
+  for (auto permutation : Permutation::ALL) {
+    auto& perm = index_.getImpl().getPermutation(permutation);
+    auto locatedTriples = LocatedTriple::locateTriplesInPermutation(
+        // TODO<qup42>: replace with the method for update block metadata once
+        // integration is done
+        idTriples, perm.metaData().blockData(), perm.keyOrder(), shouldExist,
+        cancellationHandle);
+    cancellationHandle->throwIfCancelled();
+    intermediateHandles[permutation] =
+        getLocatedTriplesPerBlock(permutation).add(locatedTriples);
+    cancellationHandle->throwIfCancelled();
+  }
+  std::vector<DeltaTriples::LocatedTripleHandles> handles{idTriples.size()};
+  for (auto permutation : Permutation::ALL) {
+    for (size_t i = 0; i < idTriples.size(); i++) {
+      handles[i].forPermutation(permutation) =
+          intermediateHandles[permutation][i];
+    }
+  }
```

Comment:
You probably should change the layout of the `handles` to be
one vector of handles per Permutation or something, then this becomes simpler here, 
but I am not yet too sure.

Reply:
This is also a `can be done after the integration as it is very local` issue.

Reply:
The handles are stored in `HashMap`s for the inserted/delted triples. They are used to resolve duplicate inserts/deletions and triples that are inserted and then deleted (or otherway around). So this would be a large change.